[c10d] Log the last enqueued and completed collective

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118582

Summary:
During debugging of some timeouted jobs, I found it difficult to
identify which rank is at fault eventhough we have logs of many ranks
reporting timeout on a specific collective seq.

If we can also report lastEqueuedSeq and lastCompletedSeq, it would be
much easier to identify,
1. whether a rank has not even join a collective call (not enqueued)
2. Or it has joined the collective call, but not completed.

For the 1st case, it is mostly likely users code problem
for the 2ed case, it could be lower-layer issues

Tags:

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang committed:
[c10d] Log the last enqueued and completed collective
Summary:
During debugging of some timeouted jobs, I found it difficult to
identify which rank is at fault eventhough we have logs of many ranks
reporting timeout on a specific collective seq.

If we can also report lastEqueuedSeq and lastCompletedSeq, it would be
much easier to identify,
1. whether a rank has not even join a collective call (not enqueued)
2. Or it has joined the collective call, but not completed.

For the 1st case, it is mostly likely users code problem
for the 2ed case, it could be lower-layer issues

Tags:

[ghstack-poisoned]

izaitsevfb commented:
Mergeability failure is legit, the conflict is with this revert: https://github.com/pytorch/pytorch/commit/fb11354594fb68d86ce3b55d39785d5c0c010fca

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1532,6 +1532,15 @@ void ProcessGroupNCCL::watchdogHandler() {
         // Report desync state in case of timeout
         if (timedOut) {
+          LOG(ERROR) << c10::str(
+              logPrefix(),
```
small suggestion- i think we already have code somewhere that prints out information about the collective which hung, when firing a timeout.  Maybe we can add the additional information to that log rather than adding a separate log line?

Note: we will likely not print anything on the one rank that is 'guilty' (it won't know locally that it timed out) - this is the classic desync debug problem)

shuqiangzhang replied:
considered adding to existing log, but the existing code that printed out the hung info is at workNCCL level, not at the processGroupNCCL level, where lastEnqueuedSeq_ and  lastCompletedSeq_ resides, So turned out that it's better print the log at processGroupNCCL level. 

Yes, for the guilty rank not knowing itself timeout is painful. If there is enough time gap between timeout and exiting, lacking such a log could be a sign of culprit.  We could also add a similar log when timeout is detected through TCPstore.

wconstab replied:
ok. well, this log is probably fine then.

shuqiangzhang committed:
Update on "[c10d] Log the last enqueued and completed collective"


Summary:
During debugging of some timeouted jobs, I found it difficult to
identify which rank is at fault eventhough we have logs of many ranks
reporting timeout on a specific collective seq.

If we can also report lastEqueuedSeq and lastCompletedSeq, it would be
much easier to identify,
1. whether a rank has not even join a collective call (not enqueued)
2. Or it has joined the collective call, but not completed.

For the 1st case, it is mostly likely users code problem
for the 2ed case, it could be lower-layer issues

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]
